### PR TITLE
Use global lock instead of per instance lock.

### PR DESCRIPTION
**Pull request by aazzolini**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30117 Use global lock instead of per instance lock.**

Multiple "trainers" could each create different instances of DistributedOptimizer, which means we can still have a race condition unless we do a trully global per worker lock.

Differential Revision: [D18603936](https://our.internmc.facebook.com/intern/diff/D18603936/)

### CODE REVIEW

File: torch/distributed/optim/optimizer.py [hunk 1]
```diff
@@ -6,17 +6,25 @@
 
 
 class _LocalOptimizer:
+    # Ideally we would only need to share a lock for instances of
+    # _LocalOptimizer that deal with the same parameters. We are
+    # making a simplifying assumption here that if there is more
+    # than one instancde of _LocalOptimizer per worker, they will
```
Comment (mrshenli) [inline torch/distributed/optim/optimizer.py]: instancde -> instance

**Review by pritamdamania87 — APPROVED** [review]

File: torch/distributed/optim/optimizer.py [hunk 2]
```diff
@@ -6,17 +6,25 @@
...
+    # be optimizing the same parameters (e.g. each data parallel
+    # trainer will create its own instance of _LocalOptimizer but
+    # they will all optimize the same parameters on each worker)
```
Comment (mrshenli) [inline torch/distributed/optim/optimizer.py]: this is a strong assumption. Is there any workaround to drop this assumption and support per parameter lock?

Comment (aazzolini) [reply torch/distributed/optim/optimizer.py]: Workarounds would be too complex and not useful for most use-cases.

**Comment by pritamdamania87** [comment]
@aazzolini @mrshenli Can we land this PR? This is a pretty serious bug.

**Comment by pietern** [comment]
The same change has been merged by @pritamdamania87 in #31404.

**Pull request closed without merging**